---
title: "Between Sample (Beta) Diversity of Microbes along a Salinity Gradient"
author: "Gurpreet Kaur"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.path = "../figures/05_CommunityAnalysis")
```

# Goals 

1. Load in phyloseq data with rooted tree.  
2. Evaluate sequencing depth and remove sample.  
3. Normalize the read counts between samples.  
4. Calculate community dissimilarities. Numbers between 0 and 1. If 0, completely similar versus if they are 1, then they're completely dissimilar.   
    a. **Sorensen**: Shared Species as a binary value: Abundance-unweighted 
    b. **Bray-Curtis**: Shared Abundant species: Abundance-weighted
    c. **(Abundance-)Weighted UNIFRAC**: Consider Abundant Species and where they fall on the tree  
5. Visualize the community data with two unconstrained Ordinations:  
    a. **PCoA**: Linear Method. Eigenvalue = how much variation is explained by each axis. Choose to view axis 1, 2, 3, etc. and plot them together.  
    b. **NMDS**: Non-linear. Smush multiple Dimensions into 2 or 3 axes. Need to report Stress value (ideally <0.15).  
6. Run statistics with PERMANOVA and betadispR.

# Setup 

## Set the seed 
```{r set-seed}
# Any number can be chosen 
set.seed(238427)
```

## Load Libraries
```{r load-libraries}
#install.packages("vegan")
pacman::p_load(tidyverse, devtools, phyloseq, patchwork, vegan,
               install = FALSE)

# Setting colors for treatments
treatment_colors <- c(
  "N0" = "brown3",
  "N120" = "dodgerblue4",
  "N180" = "darkolivegreen4",
  "N240" = "gold",
  "N360" = "deeppink4")
```



## Load Data 
```{r load-physeq}
# Load in rooted phylogenetic tree! 
load("data/03_Phylogenetic_Tree/phytree_preprocessed_physeq.RData")
midroot_physeq
unrooted_physeq
```

# Explore Read Counts 

## Raw read depth 
```{r calc-seq-depth}
# Calculate the total number of reads per sample. 
raw_TotalSeqs_df  <- 
  midroot_physeq %>%
  # calculate the sample read sums 
  sample_sums() %>%
  data.frame()
# name the column 
colnames(raw_TotalSeqs_df)[1] <- "TotalSeqs"
  
head(raw_TotalSeqs_df)

# make histogram of raw reads 
raw_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  scale_x_continuous(limits = c(0, 50000)) + 
  labs(title = "Raw Sequencing Depth Distribution") + 
  theme_classic()
```


## Remove lowly seq sample 
```{r low-seq-sample}
raw_rooted_physeq <- 
  midroot_physeq %>%
  # remove lowly seq sample that was outlier in alpha diversity analysis
  subset_samples(names != "SRR11364374") %>%
  # any asvs unique to this sample will also be removed 
  prune_taxa(taxa_sums(.) > 0, .)

# Inspect 
raw_rooted_physeq

# what is the minimum number of sequences 
raw_rooted_physeq %>%
  sample_sums() %>%
  min()
```

## Normalize read counts  
```{r scale-reads}
### scale_reads function
#################################################################################### 
# Function to scale reads: http://deneflab.github.io/MicrobeMiseq/ 
# Scales reads by 
# 1) taking proportions
# 2) multiplying by a given library size of n
# 3) rounding 
# Default for n is the minimum sample size in your library
# Default for round is floor

matround <- function(x){trunc(x+0.5)}

scale_reads <- function(physeq, n = min(sample_sums(physeq)), round = "round") {
  
  # transform counts to n
  physeq.scale <- transform_sample_counts(physeq, function(x) {(n * x/sum(x))})
  
  # Pick the rounding functions
  if (round == "floor"){
    otu_table(physeq.scale) <- floor(otu_table(physeq.scale))
  } else if (round == "round"){
    otu_table(physeq.scale) <- round(otu_table(physeq.scale))
  } else if (round == "matround"){
    otu_table(physeq.scale) <- matround(otu_table(physeq.scale))
  }
  
  # Prune taxa and return new phyloseq object
  physeq.scale <- prune_taxa(taxa_sums(physeq.scale) > 0, physeq.scale)
  return(physeq.scale)
}
```


## Scale the reads and check the distribution of the seq depth 

This is where one might decide use rarefaction to normalize data. 
```{r scale-physeq}
min(sample_sums(raw_rooted_physeq))

# Scale reads by the above function
scaled_rooted_physeq <- 
  raw_rooted_physeq %>%
  scale_reads(round = "matround")

# Calculate the read depth 
scaled_TotalSeqs_df <- 
  scaled_rooted_physeq %>%
  sample_sums() %>%
  data.frame()
colnames(scaled_TotalSeqs_df)[1] <-"TotalSeqs"

# Inspect
head(scaled_TotalSeqs_df)

# Check the range of the data 
min_seqs <- min(scaled_TotalSeqs_df$TotalSeqs); min_seqs
max_seqs <- max(scaled_TotalSeqs_df$TotalSeqs); max_seqs
# range
max_seqs - min_seqs

# Plot Histogram 
scaled_TotalSeqs_df %>%
  ggplot(aes(x = TotalSeqs)) + 
  geom_histogram(bins = 50) + 
  scale_x_continuous(limits = c(0, 50000)) + 
  labs(title = "Scaled Sequencing Depth at 17422") + 
  theme_classic()
```

# Cacluate & Visualize community dissimilarity 

Exploratory analyses from Paliy & Shankar (2016) paper, which is using unconstrained ordination methods like PCoA. 

## PCoA 

### Sorensen
```{r sorensen-PCoA}
# Calculate sorensen dissimilarity: Abundance-unweighted of shared taxa
scaled_soren_pcoa <-  
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "bray", binary = TRUE)

#str(scaled_soren_pcoa)

# Plot the ordination 
soren_treatment_pcoa <- plot_ordination(
  physeq = scaled_rooted_physeq,
  ordination = scaled_soren_pcoa,
  color = "treatment",
  shape = "year",
  title = "Sorensen PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = treatment)) +
  scale_color_manual(values = treatment_colors) + 
  theme_bw()
# Show the plot 
soren_treatment_pcoa
```

Here, I am evaluating the shared taxa by presence/absence (abundance-unweighted) in the Sorensen metric. 

- Axis 1 = ~12.2% of variation  
- Axis 2 = ~10% of variation 

This means a total of 30% variation in the data in these two axes. Axis 1 showed three distinct clusters from
1. N0 at the bottom from 1st and 5th years 
2. Microbial taxa from N120, N240 and N360 in the middle 
3. Microbial taxa from N180 in the 5th year

Axis 2 showed three distinct clusters from
1.N0
2.N120 
3.N180, N240 and N360


### Bray
```{r bray-PCoA}
# Calculate the BC distance
scaled_BC_pcoa <- 
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "bray")

# Plot the PCoA
bray_treatment_pcoa <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_BC_pcoa,
    color = "treatment",
    shape = "year",
    title = "Bray-Curtis PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = treatment)) +
  scale_color_manual(values = c(treatment_colors)) + 
  theme_bw()
bray_treatment_pcoa
```

From the Bray-curtis analysis:

- Axis 1 = ~13% of variation  
- Axis 2 = ~18% of variation 

This means that this analysis explains 31% of the variation in the data in these two axes, which is more than the previous plot with the Sorensen Dissimilarity. Abundance does seem to have an influence!!

Samples are now separated opposite to how they were separated in Sorenson.


### Weighted Unifrac
```{r wUnifrac-PCoA}
# Calculate the BC distance
scaled_wUNI_pcoa <- 
  ordinate(physeq = scaled_rooted_physeq,
    method = "PCoA",
    distance = "wunifrac")

str(scaled_wUNI_pcoa)

# Plot the PCoA
wUNI_treatment_pcoa <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_wUNI_pcoa,
    color = "treatment",
    shape = "year",
    title = "Weighted Unifrac PCoA") +
  geom_point(size=5, alpha = 0.5, aes(color = treatment)) +
  scale_color_manual(values = c(treatment_colors)) + 
  theme_bw()
wUNI_treatment_pcoa
```

Here, I am evaluating the shared taxa and then weighting them by their abundances, which provides more influence for species that are more abundant.    

- Axis 1 = ~34% of variation  
- Axis 2 = ~27% of variation 

This means we explain *61*% of the variation in the data in these two axes, which is significantly more than the previous plots with the taxonomic dissimilarity measures. This plots shows two distinct clusters in both the axes, axis 2 showing all the samples form 10th year and axis 1 showing all the samples from all treatments in 1st and 5th year.

## Combine PCoAs

Let's plot all three together into one plot to have a concise visualization of the three metrics. 
```{r pcoa-together, fig.width=8, fig.height=3.5}
(soren_treatment_pcoa + theme(legend.position = "none")) + 
  (bray_treatment_pcoa + theme(legend.position = "none")) + 
    (wUNI_treatment_pcoa + theme(legend.position = "none"))
```

## NMDS 

### Weighted Unifrac

Since we did 3 of the dissimilarity metrics for the PCoA, let's just plot one example of them for the NMDS plotting. Here, we will use weighted Unifrac 
```{r wUnifrac-NMDS}
# Calculate the Weighted Unifrac distance
scaled_wUNI_nmds <- 
  ordinate(
    physeq = scaled_rooted_physeq,
    method = "NMDS",
    distance = "wunifrac")

# Plot the PCoA
wUNI_treatment_nmds <- 
  plot_ordination(
    physeq = scaled_rooted_physeq,
    ordination = scaled_wUNI_nmds,
    color = "treatment",
    shape = "year",
    title = "Weighted Unifrac NMDS") +
  geom_point(size=5, alpha = 0.5, aes(color = treatment)) +
  scale_color_manual(values = c(treatment_colors)) + 
  theme_bw()
wUNI_treatment_nmds
```



```{r combined-wUnifrac, fig.width=6, fig.height=3.5}
(wUNI_treatment_pcoa + theme(legend.position = "none")) + 
  (wUNI_treatment_nmds + theme(legend.position = "none"))
```

PCoA shows more variability as compared to NMDS.

# Statistical Significance Testing 

## PERMANOVA 
```{r PERMANOVA}
# Calculate all three of the distance matrices
scaled_sorensen_dist <- phyloseq::distance(scaled_rooted_physeq, method = "bray", binary = TRUE)
scaled_bray_dist <- phyloseq::distance(scaled_rooted_physeq, method = "bray")
scaled_wUnifrac_dist <- phyloseq::distance(scaled_rooted_physeq, method = "wunifrac")

# make a data frame from the sample_data
# All distance matrices will be the same metadata because they 
# originate from the same phyloseq object. 
metadata <- data.frame(sample_data(scaled_rooted_physeq))

# Adonis test
# In this example we are testing the hypothesis that the five stations
# that were collected have different centroids in the ordination space 
# for each of the dissimilarity metrics, we are using a discrete variable 
adonis2(scaled_sorensen_dist ~ treatment, data = metadata)
adonis2(scaled_bray_dist ~ treatment, data = metadata)
adonis2(scaled_wUnifrac_dist ~ treatment, data = metadata)
```

We see that the most variation is explained by Bray-Curtis and Unifrac and also has the highest F-statistic.

```{r PERMANOVA-multiVariable}
# We might also care about other variables
# Here, we will add date and fraction as variables
# multiplicative model ORDER MATTERS! 
adonis2(scaled_sorensen_dist ~ treatment*year, data = metadata)
adonis2(scaled_bray_dist ~ treatment*year, data = metadata)
# Note that the ORDER MATTERS!
adonis2(scaled_wUnifrac_dist ~ treatment*year, data = metadata)
adonis2(scaled_wUnifrac_dist ~ year*treatment, data = metadata)
```

From here, we see that the highest F-statistic is observed in Unifrac and year of sample selections impacts our results. The interaction between treatment type and the year of sampling also impacts are results from microbial community analysis, which are shown further.


## BetaDispR

The PERMANOVA is sensitive to variance/dispersion in the data. Therefore, we need to run a homogeneity of dispersion test to test for the sensitivity of our PERMANOVA results to variance. 
```{r betadispR}
# Homogeneity of Disperson test with beta dispr
# Sorensen 
beta_soren_treatment <- betadisper(scaled_sorensen_dist, metadata$treatment)
permutest(beta_soren_treatment)

# Bray-curtis 
beta_bray_treatment <- betadisper(scaled_bray_dist, metadata$treatment)
permutest(beta_bray_treatment)

# Weighted Unifrac 
beta_bray_treatment <- betadisper(scaled_wUnifrac_dist, metadata$treatment)
permutest(beta_bray_treatment)
```

Here, we see that our variance if impacted by the treatment type.

# Taxonomic Composition

## Phylum
```{r phylum-colors}
# Set the phylum colors
phylum_colors <- c(
  Acidobacteriota = "navy", 
  Actinobacteriota = "darkslategray2", 
  Armatimonadota = "deeppink1",
  Alphaproteobacteria = "plum2", 
  Bacteroidota = "gold", 
  Bacteroidetes = "turquoise",
  Betaproteobacteria = "plum1", 
  Bdellovibrionota = "red1",
  Chloroflexi="black", 
  Crenarchaeota = "firebrick",
  Cyanobacteria = "limegreen",
  Deltaproteobacteria = "grey", 
  Desulfobacterota="magenta",
  Firmicutes = "#3E9B96",
  Gammaproteobacteria = "greenyellow",
  Gammatimonadetes = "khaki1",
  Myxococcota = "#B5D6AA",
  Nitrospirota = "palevioletred1",
  Nitrospirae = "lavender",
  Proteobacteria = "royalblue",
  Planctomycetota = "darkorange",
  Patescibacteria = "yellow",
  Thermoplasmatota = "green",
  Verrucomicrobiota = "darkorchid1",
  other = "grey")
```

```{r phylum-composition}
# Calculate the phylum relative abundance 
# Note: The read depth MUST be normalized in some way: scale_reads
phylum_df <- 
  scaled_rooted_physeq %>%
  # agglomerate at the phylum level 
  tax_glom(taxrank = "Phylum") %>% 
  # Transform counts to relative abundance 
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format 
  psmelt() %>%
  # Filter out Phyla that are <1% - get rid of low abundant Phyla
  dplyr::filter(Abundance > 0.01) %>%
  # fix the order of date
  mutate(date = fct_relevel(year, c("1", "5", "10")),
         station = fct_relevel(treatment, c("N0", "N120",
                                          "N180", "N240",
                                          "N360")))


# Stacked Bar Plot With All phyla 
# Plot Phylum Abundances - make sure to load phylum_colors 
phylum_df %>%
  ggplot(aes(x = names, y = Abundance, fill = Phylum)) + 
  facet_wrap(year~treatment, scales = "free_x", ncol = 5) + 
  geom_bar(stat = "identity", color = "black") + 
  labs(title = "Phylum Composition") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_blank())

## Make each phyla it's own row 
phylum_df %>%
  # Warning: It's important to have one sample per x value, 
  # otherwise, it will take the sum between multiple samples
  ggplot(aes(x = treatment, y = Abundance, fill = Phylum)) + 
  facet_grid(Phylum~year, scale = "free") + 
  geom_bar(stat = "identity") + 
  labs(title = "Phylum Composition") + 
  scale_fill_manual(values = phylum_colors) + 
  theme_bw() + 
  theme(axis.text.x = element_text( hjust = 1, vjust = 1, size = 0.5),
        strip.text.y = element_text(angle = 360,  size = 10 )) 
```

```{r phylum-treatment-actinobacteria, fig.width=5, fig.height=3.5}
# Narrow in on a specific group
# Actinobacteriota - y: abundance, x: treatment, dot plot + boxplot
phylum_df %>%
  dplyr::filter(Phylum == "Actinobacteriota") %>%
  # build the plot 
  ggplot(aes(x = treatment, y = Abundance, 
             fill = treatment, color = treatment)) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  facet_wrap(year~treatment, scales = "free_x", ncol = 5) +
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinobacteriota Phylum Abundance") + 
  scale_color_manual(values = treatment_colors) + 
  scale_fill_manual(values = treatment_colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")

# Bacteroidota
phylum_df %>%
  dplyr::filter(Phylum == "Bacteroidota") %>%
  # build the plot 
  ggplot(aes(x = treatment, y = Abundance, 
             fill = treatment, color = treatment)) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  facet_wrap(year~treatment, scales = "free_x", ncol = 5) +
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Bacteroidota Phylum Abundance") + 
  scale_color_manual(values = treatment_colors) + 
  scale_fill_manual(values = treatment_colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")

# Acidobacteriota
phylum_df %>%
  dplyr::filter(Phylum == "Acidobacteriota") %>%
  # build the plot 
  ggplot(aes(x = treatment, y = Abundance, 
             fill = treatment, color = treatment)) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  facet_wrap(year~treatment, scales = "free_x", ncol = 5) +
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidobacteriota Phylum Abundance") + 
  scale_color_manual(values = treatment_colors) + 
  scale_fill_manual(values = treatment_colors) + 
    theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
          legend.position = "right")
```

## Family
```{r family-composition, fig.width=10, fig.height=4}
# Calculate the Family relative abundance 
# Note: The read depth MUST be normalized in some way: scale_reads

family_df <- scaled_rooted_physeq %>%
  # agglomerate at the phylum level 
  tax_glom(taxrank = "Family") %>% 
  # Transform counts to relative abundance 
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format 
  psmelt() %>%
  # Filter out Phyla that are <1% - get rid of low abundant Phyla
  filter(Abundance > 0.01) %>%
  mutate(date = factor(year, levels = c("1", "5", "10")),
         station = factor(treatment, levels = c("N0", "N120", "N180", "N240", "N360")))

# Check family_df
#str(family_df)

# Plot family 
# Actinobacteriota Families 
family_df %>%
  dplyr::filter(Phylum == "Actinobacteriota") %>%
  # build the plot 
  ggplot(aes(x = treatment, y = Abundance, 
             fill = treatment, color = treatment)) + 
  facet_wrap(year~treatment, scales = "free_x", ncol = 5) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinobacteriota Family Abundance") + 
  scale_color_manual(values = treatment_colors) + 
  scale_fill_manual(values = treatment_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")

```



```{r plot-cyano-families, fig.width=8, fig.height=4}
# Plot family 
family_df %>%
  dplyr::filter(Phylum == "Patescibacteria") %>%
  # build the plot 
  ggplot(aes(x = treatment, y = Abundance, 
             fill = treatment, color = treatment)) + 
  facet_wrap(.~treatment, scales = "free_x", nrow = 1) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Patescibacteria Family Abundance") + 
  scale_color_manual(values = treatment_colors) + 
  scale_fill_manual(values = treatment_colors) + 
  theme(axis.text.x = element_blank())

```


## Genus
```{r genus-plots, fig.width=12, fig.height=4}
# Calculate the Family relative abundance 
# Note: The read depth MUST be normalized in some way: scale_reads
genus_df <- 
  scaled_rooted_physeq %>%
  # agglomerate at the phylum level 
  tax_glom(taxrank = "Genus") %>% 
  # Transform counts to relative abundance 
  transform_sample_counts(function (x) {x/sum(x)}) %>%
  # Melt to a long format 
  psmelt() %>%
  # Filter out Phyla that are <1% - get rid of low abundant Phyla
  dplyr::filter(Abundance > 0.01) %>%
  # fix the order of date
  mutate(date = fct_relevel(year, c("1", "5", "10")),
         station = fct_relevel(treatment, c("N0", "N120",
                                          "N180", "N240",
                                          "N360")))


# Actinobacteriota
# Plot genus 
genus_df %>%
  dplyr::filter(Phylum == "Actinobacteriota") %>%
  # build the plot 
  ggplot(aes(x = treatment, y = Abundance, 
             fill = treatment, color = treatment)) + 
  facet_wrap(.~Genus) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Actinobacteriota Genus Abundance") + 
  scale_color_manual(values = treatment_colors) + 
  scale_fill_manual(values = treatment_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")

# Acidobacteriota
# Plot genus 
genus_df %>%
  dplyr::filter(Phylum == "Acidobacteriota") %>%
  # build the plot 
  ggplot(aes(x = treatment, y = Abundance, 
             fill = treatment, color = treatment)) + 
  facet_wrap(.~Genus) + 
  geom_boxplot(alpha = 0.5, outlier.shape = NA) + # outliers not plotted here in boxplot 
  geom_jitter() + 
  theme_bw() + 
  labs(title = "Acidobacteriota Genus Abundance") + 
  scale_color_manual(values = treatment_colors) + 
  scale_fill_manual(values = treatment_colors) + 
  theme(axis.text.x = element_text(angle = 30, hjust = 1, vjust = 1),
        legend.position = "bottom")

```

#Conclusion:
In this entire project, I learnt a lot with quality checking, preprocessing data, making phylogenetic tree and performing community analysis. From the data set, I was working on, key things that I observed was:
1. Distinction between samples collected in 1st and 5th year when compared the final year i.e; 10th.
2. The analysis performed linked back to the findings of the authors as shown by relative richness plot:
  a. Acidobacteria: Decreased in all the treatments from 1st to 10th year.
  b. Actinobacteria: Increased in all the treatments from 1st to 10th year.
  c. Bacteroidota: Decreased in all the treatments from 1st to 10th year.
  d. Patescibacteria: Decreased in all the treatments from 1st to 10th year.
  
The results show the effect of long term N fertilizer application in wheat led to decrease in most of the dominant soil bacterial species.


# Session Information 
For reproducibility 
```{r session_info}
devtools::session_info()
```
